src/fsharp/CompileOps.fs

```diff
@@ -2549,6 +2547,56 @@ let GetAutoOpenAttributes ilg ilModule =
 
 let GetInternalsVisibleToAttributes ilg ilModule = 
     ilModule |> GetCustomAttributesOfILModule |> List.choose (TryFindInternalsVisibleToAttr ilg)
+
+let findPrimaryAssembly (data: TcConfigBuilder) =
+    let assemblies =
+        data.referencedDLLs
+        |> List.map (fun r -> r.Text)
+
+    let readerSettings: ILReaderOptions = 
+        { pdbDirPath=None
+          ilGlobals = EcmaMscorlibILGlobals
+          reduceMemoryUsage = data.reduceMemoryUsage
+          metadataOnly = MetadataOnlyFlag.Yes
+          tryGetMetadataSnapshot = data.tryGetMetadataSnapshot }
+
+    let nameSystem = ["System"]
+    let nameObject = "Object"
+
+    let hasNoPiaLocalTypes (reader: ILModuleReader) =
+        reader.ILModuleDef.TypeDefs.AsArray
+        |> Array.exists (fun (ilTypeDef: ILTypeDef) ->
+            ilTypeDef.CustomAttrs.AsArray
+            |> Array.exists (fun attr -> attr.Method.MethodRef.DeclaringTypeRef.Name = "System.Runtime.InteropServices.TypeIdentifierAttribute"))
+
+    let hasBaselessObjectClass (reader: ILModuleReader) =
+        reader.ILModuleDef.TypeDefs.AsArrayOfPreTypeDefs
+        |> Array.exists (fun (ilPreTypeDef: ILPreTypeDef) ->
+            if ilPreTypeDef.Name = nameObject && ilPreTypeDef.Namespace = nameSystem then
+                let ilTypeDef = ilPreTypeDef.GetTypeDef()
+                ilTypeDef.IsClass && ilTypeDef.Implements.IsEmpty
+            else
+                false)
+
+    let isCandidate (reader: ILModuleReader) =
+        reader.ILAssemblyRefs.IsEmpty &&
+        hasBaselessObjectClass reader &&
+        not (hasNoPiaLocalTypes reader)
+
+    let candidates =
+        assemblies
+        |> List.choose (fun path ->
+            let reader = OpenILModuleReader path readerSettings
+            if isCandidate reader && reader.ILModuleDef.HasManifest then
+                mkRefToILAssembly reader.ILModuleDef.ManifestOfAssembly
+                |> Some
+            else
+                None)
+
+    match candidates with
+    | [] -> error(InternalError("Primary assembly not found.", Range.range0))
+    | [candidate] -> candidate
+    | _ -> error(InternalError("Two or more possible primary assemblies were found. Consider fixing the assembly references.", Range.range0))
```

Comment:
Are we sure this is an internal error and never surfaces? "consider fixing" implies this may be seen by the user. Also: how?

Reply:
This is not an internal error. I have this as a placeholder for now and will be changed to a more proper error by adding one in FSComp.txt.